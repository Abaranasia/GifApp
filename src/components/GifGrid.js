import React from 'react'
import GifGridItem from './GifGridItem';
import useFetchGifs from './hooks/useFetchGifs';

export default function GifGrid({categoria}) {
/*** 
 * 
 * Componente que obtiene las imágenes de una categoría llamando a useFetchGifs 
 * y las presenta en la capa .resultados empleando el componente GifGridItem
 *  
 * ***/ 


   const {data:images, loading} = useFetchGifs(categoria);
   // Renombramos data a images para facilitar la lectura del código

    return (
        <>
            <h2>{categoria}</h2>
            
            {loading && <p>Loading...</p>}

            <div className="resultados animate__animated animate__backInDown">
               { images.map(image => (
                  <GifGridItem key={image.id} {...image}/>
               ))
                }
            </div>
        </>
    )
}
