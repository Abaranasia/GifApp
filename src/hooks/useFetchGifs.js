import {useState, useEffect} from 'react'
import { getGifs } from '../components/GetGifs';

export default function useFetchGifs(categoria) {
//export const useFetchGifs = ( category ) => // esta declaración implica la llamada con { }

/*** 
 * 
 * Hook que gestiona la descarga de nuevas imágenes de una categoría desde la api con getGifs
 * Emplea un estado con las imágenes y un booleano para mensaje de "loading.."
 * Actualiza el mensaje "loading.." a false cuando se han recibido las imágenes desde getGifs
 * Devuelve el estado con las fotos y "loading.."
 *  
 * ***/ 

    const [state, setState] = useState({
        data: [],
        loading: true
        });

        useEffect(() => {
            getGifs(categoria)
            .then(fotos => {
                setState({
                    data: fotos,
                    loading: false
                })
            });
            
        }, [categoria]) // dato dependiente; useEffect se ejecuta de nuevo si este valor cambia 

        
    return state 
}
