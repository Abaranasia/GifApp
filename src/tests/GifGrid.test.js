import React from "react";
import '@testing-library/jest-dom';

import { shallow } from "enzyme";
import GifGrid from '../components/GifGrid';
import useFetchGifs from '../hooks/useFetchGifs';

jest.mock('../hooks/useFetchGifs');


describe('Pruebas GifGrid', () => {

    const categoria="Gantz";
    

    test('Debe coincidir con el snapshot', () => {

        useFetchGifs.mockReturnValue({ // Simula la devoluci贸n de un objeto por parte del hook, vaci贸 en este caso
            data: [],
            loading: true
        });
    
        const wrapper= shallow(<GifGrid categoria= { categoria } />);

        expect(wrapper).toMatchSnapshot();
    })
    


    test('Debe mostrar items cuando recibimos una respuesta de GetGifs', () => {
        
        const respuesta = [ //array con 1 elemento para la simulaci贸n
            {
                id: 1,
                title: "algo",
                url: "http://localhost/foto.jpg"
            }]

        useFetchGifs.mockReturnValue({ // Simula la devoluci贸n de un objeto por parte del hook, con datos
            data: [respuesta],
            loading: false
        });


        const wrapper= shallow(<GifGrid categoria= { categoria } />);

        expect(wrapper).toMatchSnapshot(); //Coincide con el snapshot
        expect (wrapper.find('p').exists() ).toBe(false) // comprobar que no se imprime el <p> Loading...</p>
        expect (wrapper.find('GifGridItem').length ).toBe(respuesta.length) //comprobar que coincide la longitud del objeto recibido
    })
    
});