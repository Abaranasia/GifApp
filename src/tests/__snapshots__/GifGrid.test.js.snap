// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas GifGrid Debe coincidir con el snapshot 1`] = `
<Fragment>
  <h2>
    Gantz
  </h2>
  <p>
    Loading...
  </p>
  <div
    className="resultados animate__animated animate__backInDown"
  />
</Fragment>
`;

exports[`Pruebas GifGrid Debe mostrar items cuando recibimos una respuesta de GetGifs 1`] = `
<Fragment>
  <h2>
    Gantz
  </h2>
  <div
    className="resultados animate__animated animate__backInDown"
  >
    <GifGridItem
      0={
        Object {
          "id": 1,
          "title": "algo",
          "url": "http://localhost/foto.jpg",
        }
      }
    />
  </div>
</Fragment>
`;
